src/libraries/Native/Unix/System.Security.Cryptography.Native/android_security.c

```diff
@@ -0,0 +1,196 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "pal_err.h"
+#include "pal_types.h"
+#include "pal_utilities.h"
+#include "pal_safecrt.h"
+
+#include "android_security.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <jni.h>
+#include <android/log.h>
+
+static JavaVM *gJvm;
+
+PALEXPORT JNIEXPORT jint JNICALL
+JNI_OnLoad(JavaVM *vm, void *reserved)
```

Comment:
One possibility to consider, to avoid explicit locks:

I've been mentioning ~everywhere that `jclass` and `jmethodID` values should be cached into global variables.

If you want to avoid locking -- and who doesn't? ;-) -- then what you can do is _move_ the `JNIEnv::FindClass()` and `JNIEnv::GetMethodIDI()` invocations for *all* types to *here*.  `JNI_OnLoad()` will *not* be invoked more than once; consequently, any "must be thread safe" init code can be done *here*:

```c
static jclass    SecureRandom_class;
static jmethodID SecureRandom_ctor;
static jmethodID SecureRandom_nextBytes;
// …

PALEXPORT JNIEXPORT jint JNICALL
JNI_OnLoad(JavaVM *vm, void *reserved)
{
    gJvm = vm;

    JNIEnv *env = GetJniEnv();
    SecureRandom_class = ToGref((*env)->FindClass(env, "java/security/SecureRandom"));
    assert(SecureRandom_class != NULL);
    SecureRandom_ctor = (*env)->GetMethodID(env, randClass, "<init>", "()V"); 
    assert(SecureRandom_ctor != NULL);
    SecureRandom_nextBytes = (*env)->GetMethodID(env, randClass, "nextBytes", "([B)V");
    assert(SecureRandom_nextBytes != NULL);
    // …

    return JNI_VERSION_1_6;
}
```